packages/entity-server/src/routes/hierarchy/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../models';
+import {
+  HierarchyRequest,
+  HierarchyRequestParams,
+  HierarchyRequestBody,
+  HierarchyRequestQuery,
+  EntityResponseObject,
+} from './types';
+
+export type RelationsRequest = HierarchyRequest<
+  HierarchyRequestParams,
+  | (EntityResponseObject & { descendants: EntityResponseObject[] })[] // groupBy: ancestor, flattenToCode: false
+  | (EntityResponseObject & { ancestor: EntityResponseObject })[] // groupBy: descendant, flattenToCode: false
+  | Record<string, string[]> // groupBy: ancestor, flattenToCode: true
+  | Record<string, string>, // groupBy: descendant, flattenToCode: true
+  HierarchyRequestBody,
+  HierarchyRequestQuery & {
+    ancestorType?: string;
+    descendantType?: string;
+    groupBy?: 'ancestor' | 'descendant';
+    flattenToCode?: 'true' | 'false';
+  }
+>;
+
+export class EntityRelationsRoute extends Route<RelationsRequest> {
+  async buildDescendantAncestorPairs(descendants: EntityType[], ancestorType: string) {
+    const { hierarchyId } = this.req.ctx;
+    const descendantAncestorMapping = await this.req.models.entity.fetchAncestorDetailsByDescendantCode(
+      descendants.map(descendant => descendant.code),
+      hierarchyId,
+      ancestorType,
+    );
+
+    return Object.entries(descendantAncestorMapping).map(([descendant, { code: ancestor }]) => ({
+      descendant,
+      ancestor,
+    }));
+  }
+
+  async buildFormattedEntitiesByCode(ancestors: EntityType[], descendants: EntityType[]) {
+    const allEntities = [...ancestors, ...descendants];
+    const formattedEntities = await this.req.ctx.formatEntitiesForResponse(allEntities);
+    const formattedEntitiesByCode: Record<string, EntityResponseObject> = {};
+    allEntities.forEach((unformattedEntity, index) => {
+      formattedEntitiesByCode[unformattedEntity.code] = formattedEntities[index];
+    });
+    return formattedEntitiesByCode;
+  }
+
+  async groupByAncestorResponse(
+    descendants: EntityType[],
+    ancestorType: string,
+    flattenToCode: boolean,
+  ) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+
+    const ancestorCodeToDescendantCodes = reduceToArrayDictionary(
+      await this.buildDescendantAncestorPairs(descendants, ancestorType),
+      'ancestor',
+      'descendant',
+    );
+
+    if (flattenToCode) {
+      return ancestorCodeToDescendantCodes;
+    }
```

Comment:
Arguably the flattenToCode logic could be used in other endpoints as well, but I don't think it's as relevant for them... we can add it if it turns out to be useful 👍